### PR TITLE
Add Upgrade guide for 5.5

### PR DESCRIPTION
**Pull request by soebbing**
Please read once and check in particular if the topics you yourself worked on are described correctly.

### CODE REVIEW

File: source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md [hunk 1]
```diff
@@ -17,14 +17,205 @@ including minor and bugfix releases, refer to the `UPGRADE.md` file found in you
...
 
+## Shopware 5.5
+
+### System requirements changes
+
+The minimum PHP version still is **PHP 5.6.4 or higher**, though **PHP 7.x is highly encouraged**. Please be aware that
+PHP 5.6 and PHP 7.0 reach their "end of life" (the end of the offical support by the PHP Group) by the end of the year.
+For performance and compatibility reasons, we recommend using PHP 7.2 if no Ioncube encrypted plugin are installed.
+
+### Removed deprecations
```
Comment (bcremer) [inline source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md]: Wording: Removal of deprecated code

Comment (soebbing) [reply source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md]: Thanks @bcremer! 💙

File: source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md [hunk 2]
```diff
@@ -17,14 +17,205 @@ including minor and bugfix releases, refer to the `UPGRADE.md` file found in you
...
+$filesystem->write($file, $contents);
+```
+
+Plugins can access their own filesystems which are instantiated and available automatically:
+    * `plugin_name.filesystem.public`
+    * `plugin_name.filesystem.private`
+
+### Sitemap
+
+To support more than 50.000 URLs that are allowed in one `sitemap.xml` file, Shopware now adds a `shopware_index.xml`,
```
Comment (shyim) [inline source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md]: sitemap_index.xml

File: source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md [hunk 3]
```diff
@@ -17,14 +17,205 @@ including minor and bugfix releases, refer to the `UPGRADE.md` file found in you
...
+### System requirements changes
+
+The minimum PHP version still is **PHP 5.6.4 or higher**, though **PHP 7.x is highly encouraged**. Please be aware that
+PHP 5.6 and PHP 7.0 reach their "end of life" (the end of the offical support by the PHP Group) by the end of the year.
+For performance and compatibility reasons, we recommend using PHP 7.2 if no Ioncube encrypted plugin are installed.
+
+### Removal of deprecated code
+
+Shopware 5.5 removes a lot of old, deprecated code. If you're updating from Shopware < v5.4.5, you might want to consider
+updating on the latest version of v5.4 first: versions v5.4.5 onwards‚ show deprecation warnings in development mode if a
```
Comment (shyim) [inline source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md]: Maybe hint the logger level configuration that comes with 5.4.6

File: source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md [hunk 4]
```diff
@@ -17,14 +17,205 @@ including minor and bugfix releases, refer to the `UPGRADE.md` file found in you
...
+
+- The methods `ModelManager::addAttribute` and `ModelManager::removeAttribute` were removed. Use `\Shopware\Bundle\AttributeBundle\Service\CrudService::update` instead.
+- The class `Shopware\Bundle\EmotionBundle\ComponentHandler\EventComponentHandler` was removed. Implement a `\Shopware\Bundle\EmotionBundle\ComponentHandler` instead.
+- The legacy models `Shopware\Models\Customer\Billing` and `Shopware\Models\Customer\Shipping` aren't available anymore.
+The same is true for their references in the `Customer` model (customer.billing, customer.shipping) and the repository
+`Shopware\Models\Customer\BillingRepository`. The matching tables `s_user_billingaddress`, `s_user_shippingaddress`
+and their attribute tables `s_user_billingaddress_attributes` and `s_user_shippingaddress_attributes` won't be removed
+on upgrade until Shopware 5.6, but they won't be kept in sync anymore.
+- The following methods got removed from `Shopware\Models\Payment\Payment`:
+     * `setClass`
```
Comment (shyim) [inline source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md]: That seems to be wrong :). They are still useable https://github.com/shopware/shopware/blob/5.5/engine/Shopware/Models/Payment/Payment.php#L611

File: source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md [hunk 5]
```diff
@@ -17,14 +17,205 @@ including minor and bugfix releases, refer to the `UPGRADE.md` file found in you
...
+containing items with e.g. 19% and 7% tax. This feature needs to be activated in *Settings*, *Checkout*, *Proportional calculation of tax positions*.
+For the proportional tax calculation to work with vouchers and modes of dispatch, be sure to set the mode of tax calculation to "auto detection" in their settings
+
+The following were made to implement this feature:
+
+* Added `Shopware\Components\Cart\ProportionalTaxCalculator` to calculate proportional taxes for the cart items
+* Added `Shopware\Components\Cart\BasketHelper` to to add items to the cart that need to be calculation in a proportional way
+* Added `Shopware\Components\Cart\ProportionalCartMerger` to merge proportional cart items into one cart item
+* Added new filter event to modify proportional vouchers `Shopware_Modules_Basket_AddVoucher_VoucherPrices`   
+* Added new column `invoice_shipping_tax_rate` to s_order, to save exact dispatch shipping tax rate.
```
Comment (shyim) [inline source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md]: new column `is_proportional_calculation` to s_order, which defines that the order is made with proportional items 😊

Comment (shyim) [inline source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md]: Maybe the hint that the shipping tax rate will be calculated approximate, when the field is null like before upgrading.

File: source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md [hunk 6]
```diff
@@ -17,14 +17,205 @@ including minor and bugfix releases, refer to the `UPGRADE.md` file found in you
...
+
+### System requirements changes
+
+The minimum PHP version still is **PHP 5.6.4 or higher**, though **PHP 7.x is highly encouraged**. Please be aware that
+PHP 5.6 and PHP 7.0 reach their "end of life" (the end of the offical support by the PHP Group) by the end of the year.
+For performance and compatibility reasons, we recommend using PHP 7.2 if no Ioncube encrypted plugin are installed.
+
+### Removal of deprecated code
+
+Shopware 5.5 removes a lot of old, deprecated code. If you're updating from Shopware < v5.4.5, you might want to consider
```
Comment (mitelg) [inline source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md]: < 5.4.6

File: source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md [hunk 7]
```diff
@@ -17,14 +17,205 @@ including minor and bugfix releases, refer to the `UPGRADE.md` file found in you
...
+* Added struct `Shopware\Components\Cart\Struct\CartItemStruct` to represent items in the cart during calculation
+* Added public function `sBasket::updateCartItems` to provide a new way of interacting with cart updates
+
+### Proportional tax calculation
+
+The proportional tax calculation allows to calculate multiple taxes for all fees, vouchers, discounts etc in baskets
+containing items with e.g. 19% and 7% tax. This feature needs to be activated in *Settings*, *Checkout*, *Proportional calculation of tax positions*.
+For the proportional tax calculation to work with vouchers and modes of dispatch, be sure to set the mode of tax calculation to "auto detection" in their settings
+
+The following were made to implement this feature:
```
Comment (mitelg) [inline source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md]: the following _what_?

File: source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md [hunk 8]
```diff
@@ -17,14 +17,205 @@ including minor and bugfix releases, refer to the `UPGRADE.md` file found in you
...
+```php
+$fileName = 'HelloWorld.txt';
+$contents = 'Uploaded by Shopware using FlySystem';
+
+$filesystem = $this->container->get('shopware.filesystem.private');
+if ($filesystem->has($fileName)) {
+    $filesystem->delete($fileName);
+}
+
+$filesystem->write($file, $contents);
```
Comment (mitelg) [inline source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md]: `$file` should be named `$fileName` right?

File: source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md [hunk 9]
```diff
@@ -17,14 +17,205 @@ including minor and bugfix releases, refer to the `UPGRADE.md` file found in you
...
+which itself contains links to one or more `sitemap.xml`. The sitemap files can be created by cronjob, live or using
+the command `sw:generate:sitemap`.
+
+If you need to add some links to the sitemap, you can simply create a service implementing the interface
+`\Shopware\Bundle\SitemapBundle\UrlProviderInterface` and give it the DIC tag `sitemap_url_provider`. The sitemap exporter
+will collect your service using the tag and export the URLs it provides along with the others.
+
+### Dynamic cache invalidation
+
+interface `Shopware\Components\HttpCache\CacheTimeServiceInterface`, tag `invalidation_date_provider`
```
Comment (mitelg) [inline source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md]: nothing more to tell about that? :disappointed:

File: source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md [hunk 10]
```diff
@@ -17,14 +17,205 @@ including minor and bugfix releases, refer to the `UPGRADE.md` file found in you
...
+
+The following were made to implement this feature:
+
+* Added `Shopware\Components\Cart\ProportionalTaxCalculator` to calculate proportional taxes for the cart items
+* Added `Shopware\Components\Cart\BasketHelper` to to add items to the cart that need to be calculation in a proportional way
+* Added `Shopware\Components\Cart\ProportionalCartMerger` to merge proportional cart items into one cart item
+* Added new filter event to modify proportional vouchers `Shopware_Modules_Basket_AddVoucher_VoucherPrices`   
+* Added new column `invoice_shipping_tax_rate` to s_order, to save exact dispatch shipping tax rate.
+
+### elasticsearch
```
Comment (janbuecker) [inline source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md]: **E**lasticsearch, it's a product name

File: source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md [hunk 11]
```diff
@@ -17,14 +17,205 @@ including minor and bugfix releases, refer to the `UPGRADE.md` file found in you
...
+each of these hooks having the opportunity to call `executeParent()`. Hence, if this happened, the original (parent) 
+implementation got called multiple times, leading to unexpected behaviours. 
+
+In Shopware 5.5 the execution model was changed to a decorative type of implementation: If more than one `replace` hook
+exists for a function, calling `executeParent()` inside the hook will execute the next `replace` hook of function. Only
+the last `replace` hook will then call the original implementation on `executeParent()`.
+ 
+### Filesystem abstraction layer
+
+As of Shopware v5.5, the use of direct file system access methods (like e.g. `fopen()`, `file()`, `is_readable()` and many more)
```
Comment (janbuecker) [inline source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md]: we don't add the version prefix, please use `5.5`

File: source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md [hunk 12]
```diff
@@ -17,14 +17,205 @@ including minor and bugfix releases, refer to the `UPGRADE.md` file found in you
...
+is being discouraged. Instead, we rely on the library <a href="https://github.com/thephpleague/flysystem" target="_blank">FlySystem</a>
+as a generic file system abstraction layer. This allows a lot of flexibility in regards to where files are being stored.
+E.g. Documents, ESD and Sitemap files can now also served from S3 or Google Cloud. We added support for 
+Amazon Web Services and Google Cloud Platform for the moment but this layer will help us in the future to support more 
+cloud services. 
+
+Multiple instances of this abstraction classes are available to plugins. The services with the ids `shopware.filesystem.public`
```
Comment (janbuecker) [inline source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md]: What do you mean with this? 
> Multiple instances of this abstraction classes are available to plugins.

File: source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md [hunk 13]
```diff
@@ -17,14 +17,205 @@ including minor and bugfix releases, refer to the `UPGRADE.md` file found in you
...
+and `shopware.filesystem.private` are available to Shopware itself and to all plugins that are active. These services are
+used to access a kind of shared "folder" in which files can be stored that might be shared between Shopware and plugins.
```
Comment (janbuecker) [inline source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md]: get rid of "some kind of" and "might". It's built for that! ;)

File: source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md (hunk 8 above)
Comment (soebbing) [reply source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md]: Right

File: source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md (hunk 7 above)
Comment (soebbing) [reply source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md]: "Changes", thanks!

File: source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md (hunk 12 above)
Comment (soebbing) [reply source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md]: Maybe multiple services would be a better term. I meant you have a public and a private global service, and both again for your plugin.

File: source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md (hunk 11 above)
Comment (soebbing) [reply source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md]: okilidokili!

File: source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md (hunk 9 above)
Comment (soebbing) [reply source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md]: I thought the product name is "elasticsearch", i have been fooled!

File: source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md (hunk 2 above)
Comment (soebbing) [reply source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md]: Uuuh, right

File: source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md (hunk 5 above)
Comment (soebbing) [reply source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md]: Done!

Comment (soebbing) [reply source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md]: Done!

File: source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md (hunk 4 above)
Comment (soebbing) [reply source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md]: Damn, i changed the upgrade.md as well

File: source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md (hunk 6 above)
Comment (soebbing) [reply source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md]: 5.4.5 shows deprecation messages as well

File: source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md (hunk 10 above)
Comment (soebbing) [reply source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md]: I thought the product name is "elasticsearch", i have been fooled!

File: source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md (hunk 9 above)
Comment (soebbing) [reply source/developers-guide/shopware-5-upgrade-guide-for-developers/index.md]: Shit, forgot to write the text around that :D

**Comment by soebbing** [comment]
Please do another check today, i'd like to publish a first version on monday. Doesn't need to be final, but id shouldn't contain any obvious mistakes anymore at least.

**Pull request merged**